### PR TITLE
perf(starter): reduce SentinelFilter hot-path digest and security reflection

### LINKED ISSUE #7: perf: reduce hot-path overhead in SentinelFilter (SHA-256 + Spring Security resolution)
SentinelFilter runs on every non-excluded request when Sentinel is enabled. Two behaviors add per-request cost:

SHA-256 for identity: hash() calls MessageDigest.getInstance("SHA-256") for every request, allocating a new digest instance each time on the hot path.

Principal resolution via reflection: resolveIdentityHash uses reflection against SecurityContextHolder and Authentication on every request so the starter does not require spring-security-core. That keeps Security optional but adds reflection overhead when Security is present.

### PR DESCRIPTION
**Pull request by CollinKabwama**
Closes [](https://github.com/CollinKabwama/ai-sentinel/issues/7)

Reduces per-request overhead in `SentinelFilter` when Sentinel is enabled and the path is not excluded.

## Changes
- **Identity hashing:** reuse a per-thread `MessageDigest` for SHA-256 (`IdentityHasher`) instead of `MessageDigest.getInstance("SHA-256")` on every request.
- **Spring Security (optional classpath):** cache `Method` handles once in `SpringSecurityPrincipalBridge`; per-request path only invokes cached reflection (no repeated `Class.forName` / `getMethod`).

### CODE REVIEW

File: ai-sentinel-spring-boot-starter/src/main/java/io/aisentinel/autoconfigure/web/IdentityHasher.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    });
+
+    private IdentityHasher() {
+    }
+
+    static String sha256Hex(String s) {
+        byte[] input = (s != null ? s : "").getBytes(StandardCharsets.UTF_8);
+        MessageDigest md = SHA256.get();
+        md.reset();
+        return HexFormat.of().formatHex(md.digest(input));
```
Comment (busingepius) [inline ai-sentinel-spring-boot-starter/src/main/java/io/aisentinel/autoconfigure/web/IdentityHasher.java]: **`HexFormat.of()` is called on every request — this partially undoes the GC savings from the `ThreadLocal`.**

Cache it as a static constant instead:

```java
private static final HexFormat HEX = HexFormat.of();
```

Then replace the usage here with `HEX.formatHex(md.digest(input))`.

File: ai-sentinel-spring-boot-starter/src/main/java/io/aisentinel/autoconfigure/web/SpringSecurityPrincipalBridge.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+                return null;
+            }
+            try {
+                Object ctx = getContext.invoke(null);
+                if (ctx == null) {
+                    return null;
+                }
+                Object auth = getAuthentication.invoke(ctx);
+                if (auth == null || !Boolean.TRUE.equals(isAuthenticated.invoke(auth))) {
+                    return null;
```
Comment (busingepius) [inline ai-sentinel-spring-boot-starter/src/main/java/io/aisentinel/autoconfigure/web/SpringSecurityPrincipalBridge.java:79]: **Silent exception swallow with no logging makes production diagnosis difficult.**

If Spring Security is present but `resolve()` fails (e.g. a misconfigured context), the filter silently falls back to IP-only identity with no trace. At minimum a debug log here would help:

```java
} catch (ReflectiveOperationException e) {
    log.debug("SpringSecurityPrincipalBridge: failed to resolve principal, falling back to IP identity", e);
    return null;
}
```

File: ai-sentinel-spring-boot-starter/src/main/java/io/aisentinel/autoconfigure/web/SpringSecurityPrincipalBridge.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+                Class<?> securityContext = Class.forName("org.springframework.security.core.context.SecurityContext");
+                Class<?> authentication = Class.forName("org.springframework.security.core.Authentication");
+                Method getContext = holder.getMethod("getContext");
+                Method getAuthentication = securityContext.getMethod("getAuthentication");
+                Method isAuthenticated = authentication.getMethod("isAuthenticated");
+                Method getPrincipal = authentication.getMethod("getPrincipal");
+                Method getName = authentication.getMethod("getName");
+                return new Bridge(getContext, getAuthentication, isAuthenticated, getPrincipal, getName, true);
+            } catch (ClassNotFoundException | NoSuchMethodException e) {
+                return noop();
```
Comment (busingepius) [inline ai-sentinel-spring-boot-starter/src/main/java/io/aisentinel/autoconfigure/web/SpringSecurityPrincipalBridge.java]: **`ExceptionInInitializerError` is not caught here — a broken Spring Security jar would crash filter initialization.**

`LinkageError` is caught separately, but if `SecurityContextHolder` fails static initialization, a `ExceptionInInitializerError` (which extends `Error`, not `LinkageError`) would propagate and bring down the filter. Broaden the catch to cover this:

```java
} catch (ClassNotFoundException | NoSuchMethodException | LinkageError | Error e) {
    return noop();
}
```

Or simply combine into:

```java
} catch (Exception | Error e) {
    return noop();
}
```

File: ai-sentinel-spring-boot-starter/src/test/java/io/aisentinel/autoconfigure/web/IdentityHasherTest.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class IdentityHasherTest {
+
+    @Test
```
Comment (busingepius) [inline ai-sentinel-spring-boot-starter/src/test/java/io/aisentinel/autoconfigure/web/IdentityHasherTest.java:21]: **Missing test for `null` input.**

The implementation handles null (`s != null ? s : ""`), but there's no test asserting this. Since `request.getRemoteAddr()` can return null (the direct caller), this boundary should be covered:

```java
@Test
void sha256Hex_nullInputTreatedAsEmptyString() {
    assertThat(IdentityHasher.sha256Hex(null))
        .isEqualTo(IdentityHasher.sha256Hex(""));
}
```

**Review by busingepius — CHANGES_REQUESTED** [review]
Good direction overall — caching `Method` handles in `SpringSecurityPrincipalBridge` and switching to a thread-local `MessageDigest` are the right fixes for issue #7. Left inline comments on a few things to address before merging:

- **Must fix:** `HexFormat.of()` allocation on every request in `IdentityHasher` (partially undoes the perf win)
- **Must fix:** `ExceptionInInitializerError` not caught in `Bridge.load()` (can crash filter init with a bad Spring Security jar)
- **Recommended:** Add a `log.debug` in the `resolve()` catch block to aid production diagnosis
- **Recommended:** Add a null-input test to `IdentityHasherTest`

File: ai-sentinel-spring-boot-starter/src/main/java/io/aisentinel/autoconfigure/web/IdentityHasher.java (hunk 1 above)
Comment (CollinKabwama) [reply ai-sentinel-spring-boot-starter/src/main/java/io/aisentinel/autoconfigure/web/IdentityHasher.java]: Implemented in 89740ba: introduced private static final HexFormat HEX = HexFormat.of(); and switched to HEX.formatHex(...) to avoid per-request HexFormat.of() allocation.

File: ai-sentinel-spring-boot-starter/src/main/java/io/aisentinel/autoconfigure/web/SpringSecurityPrincipalBridge.java (hunk 2 above)
Comment (CollinKabwama) [reply ai-sentinel-spring-boot-starter/src/main/java/io/aisentinel/autoconfigure/web/SpringSecurityPrincipalBridge.java:79]: Addressed in 89740ba: added debug logging in resolve() when reflection fails:
SpringSecurityPrincipalBridge: failed to resolve principal, falling back to IP identity.

File: ai-sentinel-spring-boot-starter/src/test/java/io/aisentinel/autoconfigure/web/IdentityHasherTest.java (hunk 4 above)
Comment (CollinKabwama) [reply ai-sentinel-spring-boot-starter/src/test/java/io/aisentinel/autoconfigure/web/IdentityHasherTest.java:21]: Added null-input coverage in IdentityHasherTest (sha256Hex_nullInputTreatedAsEmptyString) in 89740ba

File: ai-sentinel-spring-boot-starter/src/main/java/io/aisentinel/autoconfigure/web/SpringSecurityPrincipalBridge.java (hunk 3 above)
Comment (CollinKabwama) [reply ai-sentinel-spring-boot-starter/src/main/java/io/aisentinel/autoconfigure/web/SpringSecurityPrincipalBridge.java]: Addressed in 89740ba: load() now catches ReflectiveOperationException | LinkageError | SecurityException and falls back to noop. This covers missing methods/classes, linkage/static-init failures, and reflective security restrictions.

**Review by busingepius — APPROVED** [review]
Awesome work. LGTM
